### PR TITLE
Add first version of backpropagation support.

### PR DESCRIPTION
**Pull request by szabadka**
This is still in progress / experimental, currently it is only implemented for normal gemma MQA attention layers, and no parallelism is added yet for backward pass.

Since we need to remember all activations from all layers, the forward pass was also reimplemented with a new activation data structure.

### CODE REVIEW

File: gemma/backward-inl.h [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+    }
+    tmp *= ss3;
+    for (size_t i = 0; i < model_dim; ++i) {
+      grad_x[offset + i] = ss * (1.0f + weights[i]) * v[offset + i] -
+                           tmp * x[offset + i];
+    }
+  }
+}
+
+static HWY_NOINLINE void InputEmbeddingVJP(
```
Comment (jan-wassenberg) [inline gemma/backward-inl.h:155]: Comment on what the VJP suffix means?

File: gemma/backward-inl.h [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+    const float* weights, const std::vector<int>& prompt,
+    const float scaling, const float* HWY_RESTRICT backward,
+    float* HWY_RESTRICT grad, size_t model_dim) {
+  for (size_t pos = 0; pos + 1 < prompt.size(); ++pos) {
```
Comment (jan-wassenberg) [inline gemma/backward-inl.h:159]: This form of loop raises GCC static analysis warnings. Recommend HWY_ASSERT(!prompt.empty()), then pos < prompt.size()-1.

File: gemma/backward-inl.h [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+    const float* HWY_RESTRICT b_out_gated =
+        backward.ffw_hidden_gated.data() + pos * kFFHiddenDim;
+    float* HWY_RESTRICT b_out = backward.ffw_hidden.data() + hidden_offset;
+    float* HWY_RESTRICT b_out_mul = b_out + kFFHiddenDim;
+    namespace hn = hwy::HWY_NAMESPACE;
+    using DF = hn::ScalableTag<float>;
+    using VF = hn::Vec<DF>;
+    DF df;
+    for (size_t i = 0; i < kFFHiddenDim; i += Lanes(df)) {
+      const auto y = Load(df, f_out + i);
```
Comment (jan-wassenberg) [inline gemma/backward-inl.h:201]: Are we sure we want to require alignment, or perhaps LoadU?

File: gemma/backward-inl.h [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+
+template <size_t kCols, size_t kRows>
+void MatMulVJP(const std::array<float, kRows * kCols>& weights,
+               const float* HWY_RESTRICT x,  // num_tokens * kCols
+               const float* HWY_RESTRICT v,  // num_tokens * kRows
+               size_t num_tokens,
+               std::array<float, kRows * kCols>& grad_w,
+               float* HWY_RESTRICT grad_x,  // num_tokens * kCols
+               hwy::ThreadPool& pool) {
+  memset(grad_x, 0, num_tokens * kCols * sizeof(grad_x[0]));
```
Comment (jan-wassenberg) [inline gemma/backward-inl.h:59]: hwy::ZeroBytes? Also below.

File: gemma/backward-inl.h [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+        const size_t k2offs = (pos2 * (kHeads + 2) + kHeads) * kQKVDim;
+        const float* HWY_RESTRICT f_k2 = forward.qkv.data() + k2offs;
+        float* HWY_RESTRICT b_k2 = backward.qkv.data() + k2offs;
+        MulByConstAndAdd(b_head_att[pos2], f_k2, b_q, kQKVDim);
+        MulByConstAndAdd(b_head_att[pos2], f_q, b_k2, kQKVDim);
+      }
+    }
+  }
+
+  for (int pos = 0; pos < num_tokens; ++pos) {
```
Comment (jan-wassenberg) [inline gemma/backward-inl.h:277]: size_t to prevent warning?

File: gemma/backward-inl.h [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+      });
+}
+
+static HWY_NOINLINE void CrossEntropyLossGrad(
+    const float* HWY_RESTRICT x, float* HWY_RESTRICT grad,
+    const Prompt& prompt, size_t vocab_size) {
+  const float scaling = -1.0 / std::log(2.0);
+  size_t num_tokens = prompt.tokens.size() - 1;
+  memset(grad, 0, num_tokens * vocab_size * sizeof(grad[0]));
+  for (size_t i = 0; i + 1 < prompt.tokens.size(); ++i) {
```
Comment (jan-wassenberg) [inline gemma/backward-inl.h:334]: Also i < size-1?

File: gemma/backward_scalar_test.cc [hunk 7]
```diff
@@ -0,0 +1,610 @@
...
+                        dx.data(), kHeads, kRows, kCols, kTokens);
+    TestGradient(dx, c_x, func, 1e-15, 1e-13,__LINE__);
+    TestGradient(grad, c_weights, func, 1e-15, 1e-13,__LINE__);
+  }
+}
+
+TEST(BackPropTest, RMSNormVJP) {
+  static const size_t K = 2;
+  static const size_t N = 64;
+  std::mt19937 gen(42);
```
Comment (jan-wassenberg) [inline gemma/backward_scalar_test.cc:102]: Consider seeding generators with hwy::Unpredictable1() from nanobenchmark.h, so that the outputs are not constexpr?

File: gemma/backward_test.cc [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+
+#if HWY_ONCE
+
+namespace gcpp {
+HWY_BEFORE_TEST(BackwardTest);
+HWY_EXPORT_AND_TEST_P(BackwardTest, TestMatMulVJP);
+HWY_EXPORT_AND_TEST_P(BackwardTest, TestMultiHeadMatMulVJP);
+HWY_EXPORT_AND_TEST_P(BackwardTest, TestRMSNormVJP);
+HWY_EXPORT_AND_TEST_P(BackwardTest, TestEndToEnd);
+#ifdef HWY_AFTER_TEST
```
Comment (jan-wassenberg) [inline gemma/backward_test.cc:264]: FYI it is now safe to drop the #ifdef because Gemma requires Highway 1.2 which has this.

File: gemma/common-inl.h [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+#include <array>
+#include <cmath>
+
+#include "gemma/activations.h"
+
+#include "hwy/base.h"
+#include "hwy/contrib/thread_pool/thread_pool.h"
+
+namespace gcpp {
```
Comment (jan-wassenberg) [inline gemma/common-inl.h:31]: Remove?

File: gemma/forward-inl.h [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+HWY_BEFORE_NAMESPACE();
+namespace gcpp {
+namespace HWY_NAMESPACE {
+namespace hn = hwy::HWY_NAMESPACE;
+
+template <typename ArrayT>
+void InputEmbedding(const ArrayT& weights, const std::vector<int>& prompt,
+                    const float scaling, float* HWY_RESTRICT output,
+                    size_t model_dim) {
+  for (size_t pos = 0; pos + 1 < prompt.size(); ++pos) {
```
Comment (jan-wassenberg) [inline gemma/forward-inl.h:54]: pos < size-1 :)

File: gemma/forward-inl.h [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+        weights.embedder_input_embedding, 0,
+        forward.final_norm_output.data() + pos * kModelDim, nullptr,
+        forward.logits.data() + pos * kVocabSize, pool);
+  }
+
+  for (size_t pos = 0; pos < num_tokens; ++pos) {
+    LogitsSoftCap(30.0f, forward.logits.data() + pos * kVocabSize, kVocabSize);
+  }
+
+  memcpy(forward.probs.data(), forward.logits.data(),
```
Comment (jan-wassenberg) [inline gemma/forward-inl.h:274]: hwy::CopyBytes?

File: gemma/forward.cc [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+#include "gemma/forward.h"
+
+// Compiles this file for multiple architectures via "foreach_target.h", to
+// which we pass the filename via macro 'argument'.
+#undef HWY_TARGET_INCLUDE
+#define HWY_TARGET_INCLUDE "gemma/forward.cc"  // NOLINT
+#include "hwy/foreach_target.h"        // IWYU pragma: keep
+
+#include "gemma/forward-inl.h"
```
Comment (jan-wassenberg) [inline gemma/forward.cc:24]: Here also after highway.h.

File: gemma/gemma.cc [hunk 13]
```diff
@@ -1192,6 +1092,57 @@ void GenerateImpl(GemmaImpl<TConfig>& gemma,
   }
 }
 
+template <class TConfig>
```
Comment (jan-wassenberg) [inline gemma/gemma.cc:1095]: FYI cl/639793810 simplified this quite a bit, and will require rebase/merge.

File: gemma/gemma_test.cc [hunk 14]
```diff
@@ -98,7 +98,7 @@ class GemmaTest : public ::testing::Test {
   gcpp::Gemma model;
 };
 
-TEST_F(GemmaTest, Geography) {
+TEST_F(GemmaTest, DISABLED_Geography) {
```
Comment (jan-wassenberg) [inline gemma/gemma_test.cc:101]: Should we revert these changes?

File: gemma/test_util.h [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+  if (sum0 > 1e-40) {
+    double norm_dot = sum01 / std::sqrt(sum0) / std::sqrt(sum1);
+    ASSERT_NEAR(norm_dot, 1.0, 1e-7)
+        << "line: " << line << " sum0: " << sum0  << " sum1: " << sum1
+        << " sum01: " << sum01;
+  }
+}
+
+// Compute gradient with the finite difference method in the complex plane.
+// If f : R->R is the tested function and F : C->C is its extenstion on the
```
Comment (jan-wassenberg) [inline gemma/test_util.h:110]: Typo: extension

File: gemma/test_util.h [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+// complex plane so that F is complex differentiable in x, then
+//
+//   F(x + ih) = F(x) + ih F'(x) + O(h^2) F''(x)
+//
+// which means that
+//
+//   F'(x) ~= Imag(F(x + ih)) / h
+//
+// This method is more numerically stable than the real-valued finite difference
+// method since we don't need to substract floating point numbers that are near
```
Comment (jan-wassenberg) [inline gemma/test_util.h:120]: Typo: subtract

File: gemma/weights.cc [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+  template <size_t N>
+  void operator()(const char* name, const std::array<float, N>& tensor) {
+    LogVec(name, tensor.data(), N);
+    total_weights += N;
+  }
+  size_t total_weights = 0;
+};
+
+template <typename TConfig>
+void LogWeightStats(const ByteStorageT& weights_u8) {
```
Comment (jan-wassenberg) [inline gemma/weights.cc:93]: Consider using hwy/stats.h for this?

File: gemma/weights.h [hunk 18]
```diff
@@ -0,0 +1,288 @@
...
+template <typename T, typename TConfig>
+ByteStorageT AllocateWeights(hwy::ThreadPool& pool) {
+  using TWeights = Weights<T, TConfig>;
+  ByteStorageT weights_u8 = hwy::AllocateAligned<uint8_t>(sizeof(TWeights));
+  TWeights* weights = reinterpret_cast<TWeights*>(weights_u8.get());
+  new (&weights->layer_ptrs) LayerPointers<T, TConfig>(pool);
+  return weights_u8;
+}
+
+#define CALL_TOP_FUNC1(name, member) func(name, weights1.member)
```
Comment (jan-wassenberg) [inline gemma/weights.h:130]: Let's add GEMMA_ prefix to macros?

File: gemma/weights.h [hunk 19]
```diff
@@ -0,0 +1,288 @@
...
+#undef CALL_TOP_FUNC4
+#undef CALL_LAYER_FUNC1
+#undef CALL_LAYER_FUNC2
+#undef CALL_LAYER_FUNC3
+#undef CALL_LAYER_FUNC4
+#undef CALL_ALL_LAYER_FUNC
+
+template<typename T, typename TConfig>
+void ZeroInit(Weights<T, TConfig>& w) {
+  memset(&w.embedder_input_embedding, 0, sizeof(w.embedder_input_embedding));
```
Comment (jan-wassenberg) [inline gemma/weights.h:242]: ZeroBytes?

File: gemma/weights.h [hunk 20]
```diff
@@ -0,0 +1,288 @@
...
+  memset(&w.final_norm_scale, 0, sizeof(w.final_norm_scale));
+  for (int i = 0; i < TConfig::kLayers; ++i) {
+    memset(w.GetLayer(i), 0, sizeof(*w.GetLayer(i)));
+  }
+}
+
+template<typename T, typename TConfig>
+void Copy(Weights<T, TConfig>& dst, const Weights<T, TConfig>& src) {
+  memcpy(&dst.embedder_input_embedding, &src.embedder_input_embedding,
```
Comment (jan-wassenberg) [inline gemma/weights.h:251]: hwy::CopyBytes?

File: gemma/weights.h [hunk 21]
```diff
@@ -0,0 +1,288 @@
...
+         sizeof(src.embedder_input_embedding));
+  memcpy(&dst.final_norm_scale, &src.final_norm_scale,
+         sizeof(src.final_norm_scale));
+  for (int i = 0; i < TConfig::kLayers; ++i) {
+    memcpy(dst.GetLayer(i), src.GetLayer(i), sizeof(*dst.GetLayer(i)));
+  }
+}
+
+template<typename T, typename TConfig>
+class WeightsWrapper {
```
Comment (jan-wassenberg) [inline gemma/weights.h:261]: Comment on what this is for? Something like "owns weights and undoes the type erasure of AllocateWeights"?

**Review by jan-wassenberg — CHANGES_REQUESTED** [review]
Wow! That's a lot of new code :)
Consider moving the backward/forward files to a subdirectory?

File: gemma/backward-inl.h (hunk 2 above)
Comment (szabadka) [reply gemma/backward-inl.h]: Done.

File: gemma/backward-inl.h (hunk 4 above)
Comment (szabadka) [reply gemma/backward-inl.h]: Done.

File: gemma/backward-inl.h (hunk 5 above)
Comment (szabadka) [reply gemma/backward-inl.h]: This needs to be an int because we use -pos later. Added static_cast to num_tokens

File: gemma/backward-inl.h (hunk 6 above)
Comment (szabadka) [reply gemma/backward-inl.h]: Done.

File: gemma/backward_scalar.h [hunk 22]
```diff
@@ -0,0 +1,351 @@
...
+    dy[i] *= (T(1.0) - scaled * scaled);
+  }
+}
+
+template<typename T>
+void CrossEntropyLossGrad(const T* x, T* dx, const Prompt& prompt, size_t V) {
+  T scaling = -1.0 / std::log(2.0);
+  size_t num_tokens = prompt.tokens.size() - 1;
+  memset(dx, 0, V * num_tokens * sizeof(x[0]));
+  for (size_t i = 0; i + 1 < prompt.tokens.size(); ++i) {
```
[the remark this answers is not in the document]
Comment (szabadka) [reply gemma/backward_scalar.h]: Done.

File: gemma/backward_scalar_test.cc (hunk 7 above)
Comment (szabadka) [reply gemma/backward_scalar_test.cc]: Wouldn't that cause flakyness, I did it like this so the tests always execute the same thing.

File: gemma/backward_test.cc (hunk 8 above)
Comment (szabadka) [reply gemma/backward_test.cc]: Done.

File: gemma/common-inl.h (hunk 9 above)
Comment (szabadka) [reply gemma/common-inl.h]: Done.

File: gemma/common_scalar.cc [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Compiles this file for multiple architectures via "foreach_target.h", to
+// which we pass the filename via macro 'argument'.
+#undef HWY_TARGET_INCLUDE
+#define HWY_TARGET_INCLUDE "gemma/common_scalar.cc"  // NOLINT
+#include "hwy/foreach_target.h"        // IWYU pragma: keep
+
+#include "gemma/ops.h"
```
[the remark this answers is not in the document]
Comment (szabadka) [reply gemma/common_scalar.cc]: Done.

File: gemma/forward-inl.h (hunk 10 above)
Comment (szabadka) [reply gemma/forward-inl.h]: Done.

File: gemma/forward-inl.h (hunk 11 above)
Comment (szabadka) [reply gemma/forward-inl.h]: Done.

File: gemma/gemma_test.cc (hunk 14 above)
Comment (szabadka) [reply gemma/gemma_test.cc:101]: I changed these to disabled so that ctest would skip them, since these need the external weights file. The test can still be run manually by adding --gtest_also_run_disabled_tests

File: gemma/ops.h [hunk 24]
```diff
@@ -877,7 +877,7 @@ static HWY_NOINLINE HWY_MAYBE_UNUSED void AddAbsolutePositionalEmbeddings(
 */
 
 static HWY_NOINLINE HWY_MAYBE_UNUSED void Rope(float* HWY_RESTRICT x,
-                                               size_t dim_qkv, size_t pos) {
+                                               size_t dim_qkv, int pos) {
```
[the remark this answers is not in the document]
Comment (szabadka) [reply gemma/ops.h:884]: This is so that the same function can be used in the backward pass. I added a comment.

File: gemma/test_util.h (hunk 15 above)
Comment (szabadka) [reply gemma/test_util.h]: Done.

File: gemma/weights.cc (hunk 17 above)
Comment (szabadka) [reply gemma/weights.cc:89]: Done.

File: gemma/weights.h (hunk 18 above)
Comment (szabadka) [reply gemma/weights.h]: Done.

File: gemma/weights.h (hunk 19 above)
Comment (szabadka) [reply gemma/weights.h]: Done.

File: gemma/weights.h (hunk 20 above)
Comment (szabadka) [reply gemma/weights.h]: Done.

File: gemma/weights.h (hunk 21 above)
Comment (szabadka) [reply gemma/weights.h:263]: Done.

**Comment by szabadka** [comment]
> Wow! That's a lot of new code :) Consider moving the backward/forward files to a subdirectory?

Done.

File: gemma/backward_scalar_test.cc (hunk 7 above)
Comment (jan-wassenberg) [reply gemma/backward_scalar_test.cc]: No worries, Unpredictable1 always returns 1, but in a way that the compiler does not know from the call site what the return value will be. Basically it's just not inline. So you can write 42*Unpredictable1 and get the same results as now, consistently.

File: gemma/gemma_test.cc (hunk 14 above)
Comment (jan-wassenberg) [reply gemma/gemma_test.cc:101]: hm, not sure everyone will know that they have to run with also_run..
I understand you want to run only the cross-entropy tests sometimes. Maybe a good way forward is to move those into a separate test?

**Review by jan-wassenberg — APPROVED** [review]
